mistralrs-core/src/gguf/gguf_tokenizer.rs

```diff
@@ -140,15 +183,43 @@ fn unigram_tokenizer(p: &PropsGGUF) -> Result<(Tokenizer, TokenizerKind, Vec<Str
         .build()?;
 
     // Add special tokens (bos, eos, unk):
-    let mut special_tokens = Vec::<String>::new();
-    for token_id in [bos, eos, unk] {
-        let token = p.tokens[token_id as usize].as_str();
+    let special_tokens = add_special_tokens(p, &mut tokenizer, bos, eos, Some(unk));
 
-        special_tokens.push(token.to_owned());
-        tokenizer.add_special_tokens(&[AddedToken::from(token.to_owned(), true)]);
+    Ok((tokenizer, TokenizerKind::Unigram, special_tokens))
+}
+
+fn bpe_tokenizer(p: &PropsGGUF) -> Result<(Tokenizer, TokenizerKind, AddedTokensCollection)> {
+    let merges = p
+        .merges
+        .as_ref()
+        .ok_or(anyhow::Error::msg("BPE tokenizer must include merges"))?
+        .iter()
+        .map(|merge| {
+            let split: (&str, &str) = merge
+                .splitn(2, ' ')
+                .collect_tuple()
+                .expect("Failed to convert split into 2-tuple");
+            (split.0.to_string(), split.1.to_string())
```

Comment:
This splits a string like "I like rust" into `("I", "like rust")`?

I haven't seen any examples where you'd have space in the input, but I assume this is referencing an existing implementation somewhere already 😅 (_not doubting your work, just curious_)

I came across [this article](https://martinlwx.github.io/en/the-bpe-tokenizer/) that mentions white-space splitting at the end, noting it is not suitable for languages like chinese.

Reply:
No problem, happy to explain. When you have a `merges` field, from all the examples I have looked at, they are encoded together, separated by a space because the BPE tokenizer inserts spaces, so there is no space token. The key point is that each element of the table is a merge *pair*. This code would split the merge representation `"he llo"` into the pair `("he", "llo")`.

I looked at the article you linked, and it also mentioned the pair construction of merges.